Migrates avatar base to TypeScript

spiritanand opened the pull request:
## Explanation

Currently, the AvatarBase component is built in JavaScript. This PR deprecates the JS component and creates a new TypeScript version.

Migrated all `.js` to `.tsx` or `.ts` files

- Converted objects to enum
- Updated storybook documentation as well as how to add a new icon
- Resolves #18063

## Screenshots/Screencaps

### Before

https://user-images.githubusercontent.com/53999998/230484021-3b3eae39-4f8a-46b6-bc15-0580094a6a7a.mp4

### After

No changes to the functionality of the component, stories, and controls all work. Updated docs.

https://user-images.githubusercontent.com/53999998/230586371-0dd9978e-feba-4ac4-aed9-7974f9f5f230.mp4

## Manual Testing Steps

- Go to the storybook build on this PR
- Search AvatarBase in IDE and story in storybook.
- Check docs align with enums

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be required.

spiritanand committed:
Converted file extension from .js to .tsx

spiritanand committed:
Updated README docs to match the enums. Resolved type errors

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand committed:
Updated AvatarBaseSizes enum values to string literals

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

legobeat committed:
Merge branch 'develop' into migrate_avatar_base_ts

legobeat committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand commented on ui/components/component-library/avatar-base/avatar-base.tsx:
```diff
@@ -0,0 +1,56 @@
+      alignItems={AlignItems.center}
+      borderRadius={BorderRadius.full}
+      variant={fallbackTextVariant}
+      textTransform="UPPERCASE"
```
I was unable to pass `TEXT_TRANSFORM.UPPERCASE` as the type of `textTransform` in `Text` component is:-
<img width="565" alt="image" src="https://user-images.githubusercontent.com/53999998/230968019-85c52306-88e6-46e8-8ec1-8de4a68fc9cf.png">
`text.tsx`
This is translating to `ts` only accepting `"UPPERCASE" | "LOWERCASE" | "CAPITALIZE" | undefined`

What do you suggest I do?

<img width="530" alt="image" src="https://user-images.githubusercontent.com/53999998/230967983-920c0f8f-8242-404e-a448-e81e70dda072.png">
`text.types.ts`

spiritanand replied:
@georgewrmarshall please have a look at this thread. We are dependent on the types `Text` accepts.

spiritanand committed:
Added TEXT_TRANSFORM.UPPERCASE as default value of textTransform in base file

legobeat committed:
lint fix

spiritanand committed:
Solved liniting errors in avatar-base

spiritanand committed:
Merge branch 'migrate_avatar_base_ts' of https://github.com/spiritanand/metamask-extension into migrate_avatar_base_ts
Merge
.

DDDDDanica commented on ui/components/component-library/avatar-base/avatar-base.types.ts:
```diff
@@ -0,0 +1,22 @@
+} from '../../../helpers/constants/design-system';
+
+export enum AvatarBaseSizes {
+  XS = 'xs',
+  SM = 'sm',
+  MD = 'md',
+  LG = 'lg',
+  XL = 'xl',
```
Can you change to Xs  to be consistent with what have been done in Icon?
https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/component-library/icon/icon.types.ts#L4
Thank you !

```suggestion
  Xs = 'xs',
  Sm = 'sm',
  Md = 'md',
  Lg = 'lg',
  Xl = 'xl',
```

legobeat committed:
Merge branch 'develop' into migrate_avatar_base_ts

georgewrmarshall commented on ui/components/component-library/avatar-base/avatar-base.types.ts:
```diff
@@ -0,0 +1,22 @@
+  XL = 'xl',
+}
+
+export interface AvatarBaseProps {
+  size?: AvatarBaseSizes;
+  children?: React.ReactNode;
+  backgroundColor?: BackgroundColor;
+  borderColor?: BorderColor;
+  color?: TextColor;
+  className?: string;
+}
```
Please add the type descriptions from the PropTypes here with appropriate adjustments for typescript

georgewrmarshall reviewed (commented):
Not sure if Text being in TS is a dependency of this PR?

georgewrmarshall commented on ui/components/component-library/avatar-base/avatar-base.types.ts:
```diff
@@ -0,0 +1,22 @@
+  TextColor,
+} from '../../../helpers/constants/design-system';
+
+export enum AvatarBaseSizes {
```
Should be singular
```suggestion
export enum AvatarBaseSize {
```

spiritanand commented on ui/components/component-library/avatar-base/avatar-base.tsx:
```diff
@@ -0,0 +1,57 @@
+      alignItems={AlignItems.center}
+      borderRadius={BorderRadius.full}
+      variant={fallbackTextVariant}
+      textTransform={TEXT_TRANSFORM.UPPERCASE}
```
Well, this solves the issue. But feel more of a hack to me. Shouldn't we be accepting the values of the enum rather than their keys?

spiritanand committed:
Made enum more consistent, added desc for AvatarBaseProps

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

georgewrmarshall reviewed (commented):
Looking good! Made one suggestion that I missed in last review. Also need to update snapshots 
```
yarn jest ui/components/component-library/avatar-base --updateSnapshot
```

spiritanand commented:
Updated :)

legobeat reviewed (changes requested):
edb650f88c5e975278588cb850ecf4706f00b44b removes the yarn release - I assume this is unintentional.

georgewrmarshall reviewed (commented):
Hey @spiritanand, could you please remove `.yarn/releases/yarn-3.2.4.cjs` file changes?

spiritanand committed:
Updated snapshots of AvatarBase and exported AvatarBaseProps in index.ts

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand commented:
Removed yarn file from git.

georgewrmarshall reviewed (commented):
Looks like there are still linting issues you can use `yarn lint` to debug. thanks!

spiritanand committed:
Made textTransform property accept right values in avatar-base.tsx

georgewrmarshall reviewed (commented):
Looks like you have some conflicts that need to be resolved

spiritanand committed:
Adding temporary changed extensions

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand committed:
Reverted files back to tsx and resolved conflicts

spiritanand committed:
Solved linting errors

georgewrmarshall commented on ui/components/component-library/avatar-base/avatar-base.types.ts:
```diff
@@ -0,0 +1,50 @@
+  Xl = 'xl',
+}
+
+export interface AvatarBaseProps {
```
I believe this interface should extend the `TextProps` interface

georgewrmarshall commented on ui/components/component-library/avatar-base/avatar-base.types.ts:
```diff
@@ -0,0 +1,50 @@
+   * Additional classNames to be added to the AvatarBase
+   */
+  className?: string;
+  /**
+   * AvatarBase also accepts all Text props including variant and all Box props
+   */
```
This comment can be removed
```suggestion
```

georgewrmarshall reviewed (commented):
Getting there made some more suggestions!

spiritanand committed:
AvatarBaseProps now extends TextProps

spiritanand commented:
@georgewrmarshall everything is working now, except the chrome test. How can we fix that?

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand committed:
Changing extension to resolve conflict

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand commented:
@georgewrmarshall I am getting the error
`AvatarBase not found via ../avatar-base/index.ts -> ../avatar-base/avatar-base.js  import/named`

spiritanand committed:
Reverted extensions back to tsx after resolving conflicts

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand commented:
@georgewrmarshall what is the update on my questions?

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

georgewrmarshall commented:
Hi @spiritanand, I'm not sure what your issue is from a quick glance. I'll try do some digging this week. It also looks like you have a failing unit test.

<img width="1127" alt="Screenshot 2023-04-27 at 8 45 08 AM" src="https://user-images.githubusercontent.com/8112138/234722330-00bc9c6a-9996-4a0d-af8c-d03fba13515f.png">

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand commented:
Thanks for the catch @legobeat, that was something I missed as well. The linting errors that I am getting are,
<img width="1366" alt="image" src="https://user-images.githubusercontent.com/53999998/234766799-885eb3a0-5cc0-48a8-be11-2876c4c49e3b.png">


I have fixed the issue you have pointed out above, though. Added `forwardRef` in `AvatarBase`

spiritanand committed:
Added forwardRef in AvatarBase

spiritanand commented:
It is working perfectly. Thank you, @legobeat and @georgewrmarshall 

You guys have been tremendously patient and helpful :)

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

georgewrmarshall commented on ui/components/component-library/avatar-base/README.mdx:
```diff
@@ -39,13 +39,13 @@ The text styles of the `AvatarBase` is based on the `size` prop. To override thi
-<AvatarBase size={AVATAR_BASE_SIZES.MD} />
-<AvatarBase size={AVATAR_BASE_SIZES.LG} />
-<AvatarBase size={AVATAR_BASE_SIZES.XL} />
+import { AvatarBaseSizes } from '../ui/component-library/avatar-base/avatar-base.types';
+import { AvatarBase } from '../../component-library/avatar-base';
+<AvatarBase size={AvatarBaseSizes.XS} />
+<AvatarBase size={AvatarBaseSizes.SM} />
+<AvatarBase size={AvatarBaseSizes.MD} />
+<AvatarBase size={AvatarBaseSizes.LG} />
+<AvatarBase size={AvatarBaseSizes.XL} />
```
Should use the singular
```suggestion
import { AvatarBaseSize } from '../ui/component-library/avatar-base/avatar-base.types';
import { AvatarBase } from '../../component-library/avatar-base';
<AvatarBase size={AvatarBaseSize.XS} />
<AvatarBase size={AvatarBaseSize.SM} />
<AvatarBase size={AvatarBaseSize.MD} />
<AvatarBase size={AvatarBaseSize.LG} />
<AvatarBase size={AvatarBaseSize.XL} />
```

georgewrmarshall reviewed (changes requested):
Looking great! Left 2 small suggestions

spiritanand committed:
Updated import name of AvatarBaseSize in README.mdx. Removed empty children attribute from AvatarBase storybook file

spiritanand committed:
Merge branch 'develop' into migrate_avatar_base_ts

spiritanand commented:
Done :)

brad-decker committed:
Merge branch 'develop' into migrate_avatar_base_ts

georgewrmarshall reviewed (approved):
LGTM! 💯 Thank you for you contribution @spiritanand

legobeat committed:
Merge branch 'develop' into migrate_avatar_base_ts

legobeat reviewed (approved):
LGTM
